File: 96.不同的二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=96 lang=cpp
 *
 * [96] 不同的二叉搜索树
 */
#include<vector>
#include<algorithm>
#include<limits.h>
using namespace std;
// @lc code=start
class Solution {
public:
    int numTrees(int n) {
        vector<int> dp(n + 1, 0);
        dp[0] = dp[1] = 1;
        for (int i = 2; i <= n; i++)
            for (int j = 1; j <= i; j++)
                dp[i] += dp[j - 1] * dp[i - j];
        return dp[n];
    }
};
// @lc code=end

File: 96.不同的二叉搜索树.java
/*
 * @lc app=leetcode.cn id=96 lang=java
 *
 * [96] 不同的二叉搜索树
 */

// @lc code=start
class Solution {
    int[][] dp;
    public int numTrees(int n) {
        dp = new int[n + 1][n + 1];
        return nums(1, n);
    }

    private int nums(int lo, int hi) {
        if (lo > hi)
            return 1;
        if (dp[lo][hi] != 0)
            return dp[lo][hi];
        int res = 0;
        for (int i = lo; i <= hi; i++) {
            int left = nums(lo, i - 1);
            int right = nums(i + 1, hi);
            res += left * right;
        }
        dp[lo][hi] = res;
        return res;
    }
}
// @lc code=end

